sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -843,6 +846,376 @@ def is_fast_reboot_enabled():
+        self.log_notice("Starting...")
+
+        # APPL_DB for CONFIG updates, and STATE_DB for insertion/removal
+        sel, asic_context = port_mapping.subscribe_port_update_event(['APPL_DB', 'STATE_DB'])
+        while not self.task_stopping_event.is_set():
+            # Handle port change event from main thread
+            port_mapping.handle_port_update_event(sel,
+                                                  asic_context,
+                                                  self.task_stopping_event,
+                                                  helper_logger,
+                                                  self.on_port_update_event)
+
+            if not self.isPortConfigDone:
+                continue
+
+            for lport, info in self.port_dict.items():
+                if self.task_stopping_event.is_set():
+                    break
+
+                if lport not in self.port_dict:
+                    continue
+
+                state = self.port_dict[lport].get('cmis_state', self.CMIS_STATE_UNKNOWN)
+                if state in [self.CMIS_STATE_UNKNOWN,
+                             self.CMIS_STATE_FAILED,
+                             self.CMIS_STATE_READY,
+                             self.CMIS_STATE_REMOVED]:
+                    continue
+
+                pport = int(info.get('index', "-1"))
+                speed = int(info.get('speed', "0"))
+                lanes = info.get('lanes', "").strip()
+                if pport < 0 or speed == 0 or len(lanes) < 1:
+                    continue
+
+                # Desired port speed on the host side
+                host_speed = speed
+
+                # Convert the physical lane list into a logical lanemask
+                #
+                # TODO: Add dynamic port breakout support by checking the physical lane offset
+                host_lanes = 0
+                phys_lanes = lanes.split(',')
+                for i in range(len(phys_lanes)):
+                    host_lanes |= (1 << i)
+
+                # double-check the HW presence before moving forward
+                sfp = platform_chassis.get_sfp(pport)
+                if not sfp.get_presence():
+                    self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_REMOVED
+                    continue
+
+                try:
+                    # Skip if XcvrApi is not supported
+                    api = sfp.get_xcvr_api()
+                    if api is None:
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_READY
+                        continue
+
+                    # Skip if it's not a paged memory device
+                    if api.is_flat_memory():
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_READY
+                        continue
+
+                    # Skip if it's not a QSFP-DD
+                    type = api.get_module_type_abbreviation()
+                    if (type is None) or (type not in ['QSFP-DD', 'QSFP_DD']):
```

Comment:
**sfp.sfp_type:** It is a constant initialized by the platform-specific chassis.py, hence it should be considered as the cage type, and it will not correctly reflect the true module type fetch from the transceiver
**sfp.get_module_type_abbreviation():** This is the true module type fetch from the transceiver, please note the cage of QSFP-DD is backward compatible with QSFP/QSFP+/QSFP28, it's expected and already happened in the field that some will attached QSFP28 modules to the platforms with QSFP-DD support.